pkg/kncloudevents/message_sender.go

```diff
@@ -137,27 +146,90 @@ func RetryConfigFromDeliverySpec(spec duckv1.DeliverySpec) (RetryConfig, error)
+			return retryConfig, err
+		}
+		retryConfig.Backoff = backoffFn
+	}
 
-		delay, err := period.Parse(*spec.BackoffDelay)
+	return retryConfig, nil
+}
+
+// generateBackoffFn returns a valid Backoff function based on the specified criteria.    In order for the respect429RetryAfter
+// value to have any effect the associated RetryConfig.CheckRetry function must have first allowed retry on 429 status codes.
+func generateBackoffFn(backoffPolicy *duckv1.BackoffPolicyType, backoffDelay *string, respect429RetryAfter bool) (Backoff, error) {
+
+	// Calculate Backoff Delay Duration
+	var backoffDelayDuration *time.Duration
+	if backoffDelay != nil {
+		delay, err := period.Parse(*backoffDelay)
 		if err != nil {
-			return retryConfig, fmt.Errorf("failed to parse Spec.BackoffDelay: %w", err)
+			return nil, fmt.Errorf("failed to parse Spec.BackoffDelay: %w", err)
+		}
+		duration, _ := delay.Duration()
+		backoffDelayDuration = &duration
+	}
+
+	// Create The Backoff Function
+	backoffFn := func(attemptNum int, resp *nethttp.Response) time.Duration {
+
+		// Parse Any 429 Retry-After Header Durations (Ignore Errors - Assume No Backoff)
+		var retryAfterDuration time.Duration
+		if respect429RetryAfter {
+			retryAfterDuration, _ = parse429RetryAfterDuration(resp)
 		}
 
-		delayDuration, _ := delay.Duration()
-		switch *spec.BackoffPolicy {
-		case duckv1.BackoffPolicyExponential:
-			retryConfig.Backoff = func(attemptNum int, resp *nethttp.Response) time.Duration {
-				return delayDuration * time.Duration(math.Exp2(float64(attemptNum)))
+		// Calculate The Appropriate Backoff Duration
+		var backoffDuration time.Duration
+		if backoffPolicy != nil && backoffDelayDuration != nil {
+			switch *backoffPolicy {
+			case duckv1.BackoffPolicyExponential:
+				backoffDuration = *backoffDelayDuration * time.Duration(math.Exp2(float64(attemptNum)))
+			case duckv1.BackoffPolicyLinear:
+				backoffDuration = *backoffDelayDuration * time.Duration(attemptNum)
 			}
-		case duckv1.BackoffPolicyLinear:
-			retryConfig.Backoff = func(attemptNum int, resp *nethttp.Response) time.Duration {
-				return delayDuration * time.Duration(attemptNum)
+		}
+
+		// Return The Larger Of The Two Backoff Durations
+		if retryAfterDuration > backoffDuration {
+			return retryAfterDuration
+		} else {
+			return backoffDuration
+		}
+	}
+
+	// Return the Backoff Function
+	return backoffFn, nil
+}
+
+// parseRetryAfterDuration returns a Duration expressing the amount of time
+// requested to wait by a 429 Retry-After header, or none if not present or invalid.
+// According to the spec (https://tools.ietf.org/html/rfc7231#section-7.1.3)
+// the Retry-After Header's value can be one of an HTTP-date or delay-seconds,
+// both of which are supported here.
+func parse429RetryAfterDuration(resp *nethttp.Response) (time.Duration, error) {
+
+	var retryAfterDuration time.Duration
+
+	if resp != nil && resp.StatusCode == nethttp.StatusTooManyRequests && resp.Header != nil {
```

Comment:
Should we also honor the 503? https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/503

Noticed look here: https://github.com/hashicorp/go-retryablehttp/blob/master/client.go#L474 

?

Reply:
whatever we decide, make sure to add a comment to the spec as well.

Reply:
> Should we also honor the 503? https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/503

Yeah I agree, good catch - thanks!